File: uni_modules/uts-any-rtc/utssdk/app-ios/RtcEffectsManager.swift
import Foundation
import BytePlusRTC

class RtcEffectsManager {

    static let shared = RtcEffectsManager()

    private var isEffectsInitialized: Bool = false
    private var videoEffect: ByteRTCVideoEffect?

    private init() {}

    func initEffects(licensePath: String, modelPath: String) -> Int32 {
        guard let engine = RtcEngineManager.shared.getEngine() else { return -1 }
        guard let effect = engine.getVideoEffectInterface() else { return -1 }
        let ret = effect.initCVResource(licensePath, withAlgoModelDir: modelPath)
        if ret == 0 {
            effect.enable()
            videoEffect = effect
            isEffectsInitialized = true
        }
        return ret
    }

    func isInitialized() -> Bool {
        return isEffectsInitialized
    }

    func enableBeauty(_ enable: Bool) -> Int32 {
        guard let engine = RtcEngineManager.shared.getEngine() else { return -1 }
        engine.enableEffectBeauty(enable)
        return 0
    }

    func setBeautyParam(mode: ByteRTCEffectBeautyMode, intensity: Float) -> Int32 {
        guard let engine = RtcEngineManager.shared.getEngine() else { return -1 }
        engine.setBeautyIntensity(mode, withIntensity: intensity)
        return 0
    }

    func setEffectNodes(_ nodePaths: [String]) -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        effect.setEffectNodes(nodePaths)
        return 0
    }

    func updateEffectNode(nodePath: String, key: String, value: Float) -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        effect.updateNode(nodePath, key: key, value: value)
        return 0
    }

    func setColorFilter(_ filterPath: String) -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        effect.setColorFilter(filterPath)
        return 0
    }

    func setColorFilterIntensity(_ intensity: Float) -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        effect.setColorFilterIntensity(intensity)
        return 0
    }

    func setVirtualBackground(sourcePath: String, sourceType: String) -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        let bgSource = ByteRTCVirtualBackgroundSource()
        if sourceType == "color" {
            bgSource.sourceType = .color
            bgSource.sourceColor = Int32(parseHexColor(sourcePath))
        } else {
            bgSource.sourceType = .image
            bgSource.sourcePath = sourcePath
        }
        effect.enableVirtualBackground("", with: bgSource)
        return 0
    }

    func disableVirtualBackground() -> Int32 {
        guard isEffectsInitialized, let effect = videoEffect else { return -1 }
        effect.disableVirtualBackground()
        return 0
    }

    func dispose() {
        guard isEffectsInitialized else { return }
        if let effect = videoEffect {
            effect.disableVideoEffect()
        }
        if let engine = RtcEngineManager.shared.getEngine() {
            engine.enableEffectBeauty(false)
        }
        videoEffect = nil
        isEffectsInitialized = false
    }

    private func parseHexColor(_ hex: String) -> Int {
        var hexStr = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if hexStr.hasPrefix("#") { hexStr.removeFirst() }
        var color: UInt64 = 0
        Scanner(string: hexStr).scanHexInt64(&color)
        return Int(color)
    }
}

File: uni_modules/uts-any-rtc/utssdk/app-android/RtcEffectsManager.kt
package uts.sdk.modules.utsAnyRtc

import com.ss.bytertc.engine.data.EffectBeautyMode
import com.ss.bytertc.engine.data.VirtualBackgroundSource
import com.ss.bytertc.engine.data.VirtualBackgroundSourceType
import com.ss.bytertc.engine.video.IVideoEffect
import io.dcloud.uts.UTSArray

object RtcEffectsManager {

    private var isEffectsInitialized: Boolean = false
    private var videoEffect: IVideoEffect? = null

    fun initEffects(licensePath: String, modelPath: String): Int {
        val engine = RtcEngineManager.getEngine() ?: return -1
        return try {
            val effect = engine.getVideoEffectInterface()
            if (effect == null) return -1
            val ret = effect.initCVResource(licensePath, modelPath)
            if (ret == 0) {
                effect.enableVideoEffect()
                videoEffect = effect
                isEffectsInitialized = true
            }
            ret
        } catch (e: Exception) {
            -2
        }
    }

    fun isInitialized(): Boolean {
        return isEffectsInitialized
    }

    fun enableBeauty(enable: Boolean): Int {
        val engine = RtcEngineManager.getEngine() ?: return -1
        return try {
            engine.enableEffectBeauty(enable)
            0
        } catch (e: Exception) { -1 }
    }

    fun setBeautyParam(mode: EffectBeautyMode, intensity: Float): Int {
        val engine = RtcEngineManager.getEngine() ?: return -1
        return try {
            engine.setBeautyIntensity(mode, intensity)
            0
        } catch (e: Exception) { -1 }
    }

    fun setEffectNodes(nodePaths: UTSArray<String>): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            effect.setEffectNodes(nodePaths.toMutableList())
            0
        } catch (e: Exception) { -1 }
    }

    fun updateEffectNode(nodePath: String, key: String, value: Float): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            effect.updateEffectNode(nodePath, key, value)
            0
        } catch (e: Exception) { -1 }
    }

    fun setColorFilter(filterPath: String): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            effect.setColorFilter(filterPath)
            0
        } catch (e: Exception) { -1 }
    }

    fun setColorFilterIntensity(intensity: Float): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            effect.setColorFilterIntensity(intensity)
            0
        } catch (e: Exception) { -1 }
    }

    fun setVirtualBackground(sourcePath: String, sourceType: String): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            val bgSource = VirtualBackgroundSource()
            if (sourceType == "color") {
                bgSource.sourceType = VirtualBackgroundSourceType.COLOR
                bgSource.sourceColor = android.graphics.Color.parseColor(sourcePath)
            } else {
                bgSource.sourceType = VirtualBackgroundSourceType.IMAGE
                bgSource.sourcePath = sourcePath
            }
            effect.enableVirtualBackground("", bgSource)
            0
        } catch (e: Exception) { -1 }
    }

    fun disableVirtualBackground(): Int {
        if (!isEffectsInitialized) return -1
        val effect = videoEffect ?: return -1
        return try {
            effect.disableVirtualBackground()
            0
        } catch (e: Exception) { -1 }
    }

    fun dispose() {
        if (!isEffectsInitialized) return
        try {
            val engine = RtcEngineManager.getEngine()
            engine?.enableEffectBeauty(false)
            videoEffect?.disableVideoEffect()
        } catch (_: Exception) {
        }
        videoEffect = null
        isEffectsInitialized = false
    }
}
